flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/MapToMapAndMultisetToMultisetCastRule.java

```diff
@@ -0,0 +1,200 @@
+            String returnVariable,
+            LogicalType inputLogicalType,
+            LogicalType targetLogicalType) {
+        final LogicalType innerInputKeyType;
+        final LogicalType innerInputValueType;
+
+        final LogicalType innerTargetKeyType;
+        final LogicalType innerTargetValueType;
+        if (inputLogicalType.is(LogicalTypeRoot.MULTISET)) {
+            innerInputKeyType = ((MultisetType) inputLogicalType).getElementType();
+            innerInputValueType = new IntType(false);
+            innerTargetKeyType = ((MultisetType) targetLogicalType).getElementType();
+            innerTargetValueType = new IntType(false);
+        } else {
+            innerInputKeyType = ((MapType) inputLogicalType).getKeyType();
+            innerInputValueType = ((MapType) inputLogicalType).getValueType();
+            innerTargetKeyType = ((MapType) targetLogicalType).getKeyType();
+            innerTargetValueType = ((MapType) targetLogicalType).getValueType();
+        }
+
+        final String innerTargetKeyTypeTerm = boxedTypeTermForType(innerTargetKeyType);
+        final String innerTargetValueTypeTerm = boxedTypeTermForType(innerTargetValueType);
+        final String keyArrayTerm = methodCall(inputTerm, "keyArray");
+        final String valueArrayTerm = methodCall(inputTerm, "valueArray");
+        final String size = methodCall(inputTerm, "size");
+        final String map = newName("map");
+        final String key = newName("key");
+        final String value = newName("value");
+
+        return new CastRuleUtils.CodeWriter()
+                .declStmt(className(Map.class), map, constructorCall(HashMap.class))
+                .forStmt(
+                        size,
+                        (index, codeWriter) -> {
+                            CastCodeBlock keyCodeBlock =
+                                    CastRuleProvider.generateCodeBlock(
+                                            context,
+                                            rowFieldReadAccess(
+                                                    index, keyArrayTerm, innerInputKeyType),
+                                            "false",
+                                            // Null check is done at the array access level
+                                            innerInputKeyType.copy(false),
+                                            innerTargetKeyType);
+
+                            CastCodeBlock valueCodeBlock =
+                                    CastRuleProvider.generateCodeBlock(
+                                            context,
+                                            rowFieldReadAccess(
+                                                    index, valueArrayTerm, innerInputValueType),
+                                            "false",
+                                            // Null check is done at the array access level
+                                            innerInputValueType.copy(false),
```

Comment:
@slinkydeveloper didn't we change this behavior recently and use a new method that deals with `NullType`?